adr/0002-generate-extension-metadata-in-deployment-module.adoc

```diff
@@ -0,0 +1,123 @@
+- *Running builds twice*. Because the circular dependency is a soft one, it works perfectly well to run a build to generate the deployment module's files, and then run a second build to pull those into the runtime module's metadata generation. Doing this would be a simple, non-breaking, change to release scripts. However, it is error-prone (if scripts are incorrectly run once-only, we just miss metadata), lengthens build times, breaks the Develocity by introducing hidden cross-module dependencies, and is generally Just Wrong.
+- *Doing nothing*. This is cheap, and non-breaking. But it means we miss an opportunity to take advantage of the fact that, at build time, we know a lot about what extensions can do and would like to make that information available to Quarkus users.
+- *Move (most) metadata generation to the deployment module* We can eliminate the circular dependency by not generating information which isn't particularly related to the runtime module in the runtime module. If the deployment module 'hosts' the `quarkus-extension.yaml`, it can harvest information from the extension's build steps and build items. This is architecturally clean, unlocks a range of capabilities in the future, and is highly disruptive in the present.
+
+== Proposal
+
+Moving metadata generation to the 'right' place is the architecturally correct thing to do.
+
+image::images/0002-resolution.svg[the proposed resolution]
+Source: https://excalidraw.com/#json=JMERR95gRufcmnUpWy6A_,Qd9efiR40LR3wv82CxyLKg
+
+So how do we minimise the impact?
+What needs to happen for it to work?
+
+* The registry reads `quarkus-extension.yaml` from the deployment module, not the runtime module. An alternate implementation would be to publish a separate `-metadata` artifact and the tooling could read that artifact.
+* The build of extensions' deployment module invokes a maven plugin (the existing one, or a new one)
+* (optional) Extension source code refactoring:  the template `runtime/src/META-INF/quarkus-extension.yaml` moves to `deployment`. This eliminates one of the cross-module dependencies we'd like to avoid to stay cache-friendly. We want this refactoring to be optional, so if the yaml template isn't found in the deployment module, tooling would look for it in the runtime module.
+
+The `quarkus-extension.properties` will continue to be generated in the runtime module. We should do an inventory to double-check that nothing in the `quarkus-extension.yaml` needs
+
+* to be accessed by tools outside the registry which expects it to be in the runtime JAR (IDEs?)
+* generated information which is only available to the extension plugin when it's running in the runtime module
+
+If so, the consumers will need to be updated to read the deployment module, or those informations should move to `quarkus-extension.properties` (if the consumer has no access to the deployment module).
```

Comment:
why is it quarkus-extension.properties would be recommended here over just having it in the existing yaml ?

Reply:
If we keep it in the existing yaml, we have two files with the same name, and different content. That means instead of talking about something being in `quarkus-extension.yaml` we'd have to say it's in `runtime/quarkus-extension.yaml` or `deployment/quarkus-extension.yaml`. It seems like a recipe for confusion.

Reply:
We will need different names if we want to do that. Or it's going to be hard to differentiate things for people gathering information about the extensions from the classpath.

Or the two files are different and clearly mention the relation between the two artifacts as part of the same extension.

Ideally, I would like two descriptors and both would contain the canonical extension they refer to e.g. `groupId:artifactId` of the runtime artifact.

Reply:
I literally came to the same realization overnight @gsmet.

Basically - we already do when you have  `groupid:someextension`  ...then look for quarkus-extesnion.yaml to know its an extension  then we look up `groupid:someextension-deployment`how about we follow the same pattern here? 

name it `quarkus-extension-deployment.yaml` to know its the -deployment and it has the groupdId:artifactId:version id  in it (just like quarkus-extension.yaml) has it.

Then you merge those values or simply lookup in deployment first, runtime second.

Then the lookup mechanism is deterministic, consistent and even mostly backwards compatible (as can just keep the runtime one or have it have less values if some key/value pairs are deemed more deployment relevant)

And reason I was hinting elsewhere in PR this ADR relates to the doc config annotation processor is that I can imagine similar would make sense for the config property metainfo. that there is a combined runtime + deployment that needs gathering. 

p.s. My original comment here missed some context: quarkus-extension.properties existed when we add the .yaml file and we discussed back then if we should merge/combine it but what I recall was that the .properties file had to stay and had to stay small as it was used in a very tight loop affecting reboot/start of quarkus - thus not deemed the good place to put content. @aloubyansky might remember the details and wether still relevant.

Reply:
I am not sure I see enough reasons for `quarkus-extension-deployment.yaml` atm.
I would still like to have a split between a build-related metadata used by bootstrap and the devtools related one. There will be some metadata present in both files and that's ok.
I would leave `quarkus-extension.properties` out of this. AFAIU, this ADR is about moving `quarkus-extension.yaml` to the deployment and possibly deploying it as its own artifact.

Reply:
@maxandersen , can we be more specific about who the "you" in "you merge those values" is? 

At the moment, we have two things that look at the metadata. 

- The ExtensionDescriptorMojo is a maven plugin that creates an enriched `quarkus-extension.yaml`, based on a source-controlled file
- The registry upload tooling takes that file and passes it to the registry. 

My proposal is that the ExtensionDescriptorMojo effectively does the merge, by running when it has access to more information. @maxandersen, is your proposal that we add a third 'thing' which does the merge?

Reply:
Actually, better yet, could you annotate https://github.com/quarkusio/quarkus/pull/40336/files#diff-3d9123e97ef17e972bccb4811c8a5222122bbd076824a551ed2e35bfdb8c8e74 with what your proposal is?

Reply:
> I am not sure I see enough reasons for `quarkus-extension-deployment.yaml` atm.

since we are going to have extensions that have quarkus-extension.yaml in both places the main reason is to just differentiate between them without having to look in what name is in the GAV. It is not an absolute necessity - just to be very explicit about it to avoid everyone thinking the two files are equal/same because they wont be - only by chance.

" this ADR is about moving quarkus-extension.yaml to the deployment and possibly deploying it as its own artifact."  - i respectfully disagree. this is not about *moving* quarkus-extension.yaml; it is about finding way how we can capture the info that deployment + runtime parts of extension has of info. Moving it is *one* out of multiple ways that can be done - what I 'm arguing is that moving it is more problematic/harder to do than simply merging content.

And the merging can be done in multiple ways:

A) deploying a separate artifact with things merged in place
B) generating a merged file into -deployment 
C) doing the merging in the tools that need it.

...in all three cases - the platform will need ability to override values from them anyway thus I'm (for now at least) -1 on introducing a separate artifact (i.e. option A)

B - has value as it simplifies consumption but you will still require for backwards compatible reasons that tools does parts of C as you have to look for quarkus-extension.yaml in "-deployment" first , then quarkus-extension.yaml in runtime. This is where calling the merged arficact `quarkus-extension-deployment.yaml` could be useful.

C - fully backwards compatible as tools will read quarkus-extension.yaml as done today, but also read quarkus-extesnion(-deployment).yaml to lookup the key first.